code/game/machinery/turret_control.dm

```diff
@@ -185,6 +186,27 @@
 
 	update_icon()
 
+/obj/machinery/turretid/proc/turretModes()
+	var/lethal_only = 0 // To check if there is a lethal turret in the area.
+	var/both_mode = 0
+	for (var/obj/machinery/porta_turret/aTurret in control_area)
+		// If turret only has lethal mode - lock switching modes
+		if(aTurret.egun == 0)
```

Comment:
You didn't fix this loop's reliance on the last member.

This loop **also** doesn't check if any turrets have lethal or stun active. Instead it disables them regardless of whether or not their current setting conflicts with new valid states. This is to say:
* Turret is lethal only and set to lethal.
* New turret is spawned, this proc is ran.
* Turret with lethal only gets disabled because reasons.

Actually, to be frank I'm not even sure what on earth this set up is trying to do. Here's a clearer alternative:
```dm
var/allow_both = TRUE
var/list/active_lethals = list()
for (var/obj/machinery/porta_turret/T in control_area)
	if (T.lethal)
		active_lethals += T
	if (!T.egun)
		modes = FALSE

egun = allow_both

if (!egun && !lethal) // Can no longer stun, but lethal = FALSE thus stun was active.
	lethal = TRUE   // Just swap everything to lethal.
	updateTurrets()
```

Reply:
`using lethal` variable of turret is misleading, because some turrets like `sniperifle` have lethal set to 0, I have no idea why it is so.

Reply:
One starts to wonder how even turrets work at this point.

My investigation seems to indicate that `egun` says wether or not two fire modes are available. And that `lethal` is then used to toggle between those two modes. And if neither is applicable then apparently it doesn't matter and only on/off toggles it? Which I guess is fine. You shouldn't really be creating new turrets with the current ones active, so t'is acceptable behaviour IMO.

Or you could just do:
```dm
if (!egun && !lethal) // Can no longer stun, but lethal = FALSE thus stun was active.
	lethal = TRUE   // Just swap everything to lethal.
	enabled = TRUE
	updateTurrets()
```

Though I have a feeling this would be more obtrusive. Maybe.

Reply:
`var/lethal = 0	//whether in lethal or stun mode` From the turret code. When I last messed with this code, I found an issue, where some turrets have both stun and lethal modes, some only lethal mode, but some troublemakers only had stun mode. I mean, to me. The turret code looks quite robust, it just has big abundance of variables, and naming of then is sometimes questionable. I guess carefully mass-renaming some like `egun` could clear things up a bit. Or maybe find a way to reduce the number of vars going on.

Main ones are in `portable_turret.dm`, obviously.
Gun ones are in `energy.dm` under `//vars passed to turrets` comment.
And then some scattered around weponds and such.